File: Programmers/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.cpp
#include <string>
#include <vector>
#include <unordered_set>

using namespace std;

int solution(vector<int> elements) {
    int n = elements.size();
    
    vector<int> extended(2 * n);
    for (int i = 0; i < 2 * n; i++)
        extended[i] = elements[i % n];
    
    unordered_set<int> sumSet;

    vector<int> prefix(2 * n + 1, 0);
    for (int i = 1; i <= 2 * n; i++)
        prefix[i] = prefix[i - 1] + extended[i - 1];

    for (int len = 1; len <= n; len++)
    {
        for (int start = 0; start < n; start++)
        {
            int sum = prefix[start + len] - prefix[start];
            sumSet.insert(sum);
        }
    }

    return sumSet.size();
}

File: Programmers/2/131701. 연속 부분 수열 합의 개수/연속 부분 수열 합의 개수.java
import java.util.*;

class Solution {
    public int solution(int[] elements) {
        int n = elements.length;
        int[] extended = new int[2 * n];

        for (int i = 0; i < 2 * n; i++) {
            extended[i] = elements[i % n];
        }

        Set<Integer> sumSet = new HashSet<>();

        int[] prefix = new int[2 * n + 1];
        for (int i = 1; i <= 2 * n; i++) {
            prefix[i] = prefix[i - 1] + extended[i - 1];
        }

        for (int length = 1; length <= n; length++) {
            for (int start = 0; start < n; start++) {
                int sum = prefix[start + length] - prefix[start];
                sumSet.add(sum);
            }
        }

        return sumSet.size();
    }
}